contracts/RestrictedNORI.sol

```diff
@@ -0,0 +1,1089 @@
+    return
+      block.timestamp < schedule.startTime
+        ? 0
+        : (totalSupply(scheduleId) * (block.timestamp - schedule.startTime)) /
+          rampTotalTime;
+  }
+
+  /**
+   * @dev Calculates the number of tokens that should be revoked from a given token holder and schedule based on their
+   * proportion of ownership of that schedule's tokens and the total number of tokens being revoked.
+   */
+  function _quantityToRevokePerTokenHolder(
+    uint256 totalQuantityToRevoke,
+    uint256 scheduleId,
+    address account
+  ) private view returns (uint256) {
+    Schedule storage schedule = _scheduleIdToScheduleStruct[scheduleId];
+    uint256 scheduleTrueTotal = schedule.totalClaimedAmount +
+      totalSupply(scheduleId);
+    // avoid division by 0
+    if (scheduleTrueTotal == 0) {
+      return 0;
+    }
+    uint256 balanceOfAccount = balanceOf(account, scheduleId);
+    uint256 claimedAmountForAccount = schedule.claimedAmountsByAddress[account];
+    uint256 revocableForAccount = ((claimedAmountForAccount +
+      balanceOfAccount) * (totalQuantityToRevoke)) / scheduleTrueTotal;
+
+    return revocableForAccount;
+  }
+
+  /**
+   * @dev The total amount of released tokens available at the current block timestamp for the schedule.
+   * Takes the maximum of either the calculated linearly released amount based on the schedule parameters,
+   * or the released amount floor, which is set at the current released amount whenever the balance of a
+   * schedule is decreased through revocation or withdrawal.
+   */
+  function _releasedBalanceOfSingleSchedule(uint256 scheduleId)
+    internal
+    view
+    returns (uint256)
+  {
+    Schedule storage schedule = _scheduleIdToScheduleStruct[scheduleId];
+    return
+      MathUpgradeable.max(
+        _linearReleaseAmountAvailable(scheduleId),
+        schedule.releasedAmountFloor
+      );
+  }
+
+  /**
+   * Grants `role` to `account`
+   *
+   * ### Requirements:
+   *    - SCHEDULE_CREATOR_ROLE cannot be granted to an owner of RestrictedNORI tokens
+   *
+   * [See OZ Access Control docs for more] (
+   *  https://docs.openzeppelin.com/contracts/3.x/access-control)
+   */
+  function _grantRole(bytes32 role, address account) internal virtual override {
+    if (role == SCHEDULE_CREATOR_ROLE) {
+      if (_addressToScheduleIdSet[account].length() > 0) {
+        revert RoleUnassignableToScheduleHolder({
+          account: account,
+          role: "SCHEDULE_CREATOR_ROLE"
+        });
+      }
+    }
+    super._grantRole(role, account);
+  }
+
+  // Disabled functions ===========================================
+  /**
+   * Overridden standard ERC1155.setApprovalForAll that will always revert
+   *
+   * @dev This function is not currently supported from external callers so we override it to revert.
+   */
+  function setApprovalForAll(address, bool) public pure override {
+    revert OperatorActionsNotSupported();
```

Comment:
can you expand on why this is disabled?

Reply:
so come to think of it... maybe we don't need to disable this, if we want to allow suppliers to grant approval to other addresses to transfer their tokens.  When scott and I were looking at this though we were confused because the implementation and the docs literally look as if anyone can approve another address to be an operator.  like there is no check for whether the sender of a `setApprovalForAll` transaction is indeed the owner of the tokens... maybe you can take a look at the inherited implementation and let me know if you're seeing something different because i'm definitely confused

Reply:
A check isn't needed because the function uses msg.sender as the owner argument here and so you can only set approvals for tokens owned by you

https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/token/ERC1155/ERC1155.sol#L405